In [10]:
#url = "https://docs.google.com/spreadsheets/d/1eZj7EOdGe16P5QWHSIIHhkclNAog13e7xF6fPMoXfEY/edit#gid=1897262175"
name = "main"

import pandas as pd
import urllib.parse

sheet_id = "1eZj7EOdGe16P5QWHSIIHhkclNAog13e7xF6fPMoXfEY"
sheet_name = "Buy_Insta_Accounts"
#sheet_name_encoded = urllib.parse.quote(sheet_name)
url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}"

In [14]:
fn = "Anthony MainSheet [current projects] - Buy_Insta_Accounts.tsv"
Buy_Insta_Accounts_df = pd.read_csv(fn, sep="\t")

## Load data

In [24]:
import os 
import pandas as pd

files = os.listdir("data")
files.sort()

df_women_event = pd.read_csv(os.path.join("data", files[0]), sep="\t")
df_buy_insta_accounts_add_on = pd.read_csv(os.path.join("data", files[1]), sep="\t")
df_buy_biz = pd.read_csv(os.path.join("data", files[2]), sep="\t")
df_buy_insta_accounts = pd.read_csv(os.path.join("data", files[3]), sep="\t")
df_women_event_mined = pd.read_csv(os.path.join("data", files[4]), sep="\t")

In [60]:
files = os.listdir("data")
files.sort()
files

['0.tsv', '1.tsv', '2.tsv', '3.tsv', '4.tsv']

## EDA

In [56]:
df_buy_biz

username  media_count                      full_name  \
0             ariannyceleste         6039       Arianny Celeste UFC Mama   
1                 sarahstage         1501                    Sarah Stage   
2     beauty_nurse_elizabeth         1060           Elizabeth Sommers RN   
3                  carlyrbel          493                      CARLY BEL   
4              houdaschimmel          507                 Houda Schimmel   
...                      ...          ...                            ...   
1467            houghtonmade            1  Memorable and Detailed Events   
1468               violacomm            3           Viola Communications   
1469  brainstorm.media.group            0         Brainstorm Media Group   
1470     littlestudio_beauty           33           Little Studio_Beauty   
1471           adtop1percent            0   Ann Dee | LA ENTREPRENEUR LA   

      following_count  follower_count first_name  \
0                 997         3288148    Celeste   
1                1472         2086691      Sarah   
2                 383         1504557  Elizabeth   
3                 191         1035410      Carly   
4                3041          693036      Houda   
...               ...             ...        ...   
1467              166              21     Lauren   
1468                3              20      Viola   
1469               32              20  Katherine   
1470              146              12   Laetitia   
1471               42               7        Ann   

                          external_url  engagement_percent  \
0     http://www.ariannyexclusive.com/                0.06   
1         https://linktr.ee/Sarahstage                0.01   
2     http://www.elizabethsommers.com/                0.23   
3                 http://carlybel.com/                0.14   
4                                  NaN                 NaN   
...                                ...                 ...   
1467                               NaN               10.95   
1468   http://violacommunications.com/                6.83   
1469  http://brainstormmediagroup.com/                0.00   
1470       http://www.littlestudio.us/                4.03   
1471                               NaN                0.00   

                                   email contact_phone_number  \
0             arianny@ariannyceleste.com                  NaN   
1               booksarahstage@gmail.com                  NaN   
2                    sommersrn@gmail.com                  NaN   
3                carlybelsnaps@gmail.com                  NaN   
4               infos@shimmypetstore.com                  NaN   
...                                  ...                  ...   
1467             laurenjmadrid@gmail.com         +13234008726   
1468        info@violacommunications.com                  NaN   
1469  katherine@brainstormmediagroup.com                  NaN   
1470            laetitia@littlestudio.fr                  NaN   
1471             adtop1percent@gmail.com                  NaN   

                     cityName       category  \
0     Los Angeles, California  Public figure   
1                         NaN            NaN   
2     Los Angeles, California  Health/beauty   
3                         NaN   Entrepreneur   
4     Los Angeles, California   Entrepreneur   
...                       ...            ...   
1467  Los Angeles, California            NaN   
1468           Tampa, Florida            NaN   
1469  Los Angeles, California            NaN   
1470                      NaN            NaN   
1471                      NaN   Entrepreneur   

                                              biography address_street  \
0     Raidens Mom 👶🏽🐶🐶🧿 Founder @girlfriendbox @them...            NaN   
1     Boy Mom 👶🏻👦🏻 Wife 💃🏻 & Online Fitness Coach 💪🏼...            NaN   
2     Aesthetic RN-BSN •Injection Artist 🌹 •Natural ...            NaN   
3     One of a kind ✨ • @carlybelx • shop @carlyclub...            NaN   
4        

In [3]:
df_buy_biz.columns

Index(['username', 'media_count', 'full_name', 'following_count',
       'follower_count', 'first_name', 'external_url', 'engagement_percent',
       'email', 'contact_phone_number', 'cityName', 'category', 'biography',
       'address_street', 'MostRecentPostDate', 'AvgLikes', 'AvgComments'],
      dtype='object')

In [31]:
df_women_event_mined.columns

Index(['username', 'userid', 'scraped_from', 'media_count', 'full_name',
       'following_count', 'follower_count', 'first_name', 'external_url',
       'email', 'contact_phone_number', 'cityName', 'category', 'biography',
       'address_street', 'Big 5'],
      dtype='object')

In [38]:
df_buy_insta_accounts.columns

Index(['username', 'Openness', 'Conscientiousness', 'Extroversion',
       'Agreeableness', 'Neuroticism', 'Type', 'media_count', 'full_name',
       'following_count', 'follower_count', 'first_name', 'external_url',
       'engagement_percent', 'email', 'contact_phone_number', 'cityName',
       'category', 'biography', 'address_street', 'AvgLikes', 'AvgComments'],
      dtype='object')

In [50]:
df_buy_biz[df_buy_biz['Openness'].notna()]

username Openness Conscientiousness Extroversion  \
0               goodeatzco      Low              High         High   
2               a1republic     High              High         High   
3           keychainsocial     High              High         High   
4        thegiftshoppe.nft    Vague               Low          Low   
5       officialclubsocial      Low              High         High   
..                     ...      ...               ...          ...   
138           irose.social     High              High         High   
139               nattcity     High               Low         High   
140             jondennill     High              High         High   
141            igmodels_co      Low              High        Vague   
142  skystrategicmarketing     High              High         High   

    Agreeableness Neuroticism      Type  media_count                full_name  \
0           Vague       Vague      Blog         1693       #GoodEatzCo | Andy   
2            High         Low  Business           30            A1 REPUBLIC ™   
3            High         Low  Business          214    Keychain Social, Inc.   
4           Vague       Vague  Business            9          the gift shoppe   
5           Vague       Vague  Business            6              Club Social   
..            ...         ...       ...          ...                      ...   
138           Low         Low  Business          338                    IROSE   
139           Low         Low  Personal         1129                 nathalie   
140          High         Low  Personal          124         Jonathan Dennill   
141         Vague       Vague      Blog          458    IG Models Worldwide ®   
142          High         Low  Business          427  Sky Strategic Marketing   

     following_count  ...  engagement_percent  \
0                145  ...                0.14   
2                613  ...                0.53   
3                  0  ...                 NaN   
4                  5  ...                0.57   
5                  2  ...                 NaN   
..               ...  ...                 ...   
138              422  ...                0.30   
139              686  ...                0.06   
140              733  ...                0.49   
141              158  ...                0.01   
142             2133  ...                0.18   

                               email contact_phone_number  \
0               goodeatzco@gmail.com                  NaN   
2               a1republic@yahoo.com         1.949934e+10   
3            info@keychainsocial.com                  NaN   
4        hello@fromthegiftshoppe.com                  NaN   
5           clubsocialmgmt@gmail.com                  NaN   
..                               ...                  ...   
138           sharon@irosesocial.com                  NaN   
139  nathalia@nfluenceconsulting.com                  NaN   
140            jon.dennill@gmail.com                  NaN   
141            igmodels.co@gmail.com                  NaN   
142  tanya@skystrategicmarketing.com         1.813337e+10   

                      cityName             category  \
0      Los Angeles, California              Blogger   
2    Newport Beach, California  Social Media Agency   
3      Los Angeles, California                  NaN   
4                          NaN                  NaN   
5      Los Angeles, California  Social Media Agency   
..                         ...                  ...   
138    Los Angeles, California      Product/service   
139                        NaN                Actor   
140                        NaN                  NaN   
141    Los Angeles, California  Social Media Agency   
142             Tampa, Florida                  NaN   

                                             biography  \
0    🍴Professional Food Consumer🤘🏻😛 Tiktok 🎥: @good...   
2    The #1 organic IG growth service for influence...   
3    Digital Social Media Agency • Unlock Your 

In [52]:
len(df_buy_insta_accounts_add_on)

2734

In [35]:
df_buy_insta_accounts_add_on

username Openness Conscientiousness Extroversion  \
0        patrickbreenmusic      Low              High          Low   
1     love.lindsey.wheeler     High              High         High   
2             authorkwells     High               Low         High   
3        aleksatodorovicev     High              High         High   
4         grigoruadisveley     High              High         High   
...                    ...      ...               ...          ...   
2729        danielarendonn      NaN               NaN          NaN   
2730         the_fifthguys      NaN               NaN          NaN   
2731              mskialee      NaN               NaN          NaN   
2732        beckie_farrant      NaN               NaN          NaN   
2733          mikelonghelt      NaN               NaN          NaN   

     Agreeableness Neuroticism      Type  media_count  \
0             High         Low  Personal          504   
1             High         Low  Personal         2381   
2             High         Low  Personal          129   
3              Low         Low  Personal          897   
4              Low         Low  Personal        10742   
...            ...         ...       ...          ...   
2729           NaN         NaN       NaN          408   
2730           NaN         NaN       NaN          167   
2731           NaN         NaN       NaN          236   
2732           NaN         NaN       NaN         2109   
2733           NaN         NaN       NaN          987   

                          full_name  following_count  ...  \
0                     Patrick Breen              893  ...   
1                   LINDSEY WHEELER             1332  ...   
2     Author Kamisha Wells RDKS CEO              314  ...   
3                 Todorović Aleksa              876  ...   
4       Anzhelika Grigoryadis Veley              980  ...   
...                             ...              ...  ...   
2729                 Daniela Rendon               36  ...   
2730                  The FifthGuys              420  ...   
2731          Ms. Lee/ Royalty حقية              102  ...   
2732                 Beckie Farrant             1339  ...   
2733                   mikelonghelt              539  ...   

                                           external_url engagement_percent  \
0                                                   NaN              2.793   
1                          https://www.lovelindsey.com/              1.063   
2              https://www.amazon.com/shop/authorkwells              1.804   
3                   https://linktr.ee/aleksatodorovicev              0.703   
4                   https://paysend.com/referral/01g6gi              0.470   
...                                                 ...                ...   
2729                   https://linktr.ee/danielarendonn              0.674   
2730                      https://discord.gg/CpdVpPaZXu              0.777   
2731                         https://linktr.ee/MSKIALEE              0.045   
2732  https://www.greatstuff.dupont.com/great-pantry...              0.498   
2733                             https://ffm.to/mmlqzdw              0.274   

                                 email  contact_phone_number  \
0        patrick@patrickbreenmusic.com                   NaN   
1            lindseyslularoe@gmail.com          +17576206307   
2               kamishawells@yahoo.com          +15627350110   
3            todorovicaskela@gmail.com                   NaN   
4     anzhelika-grigorjadis@rambler.ru                   NaN   
...                                ...                   ...   
2729         daniela@danielarendon.com          +17868997656   
2730    thefifthguysofficial@gmail.com                   NaN   
2731              klee@keybridgeus.com          +17706424307   
2732     infarrantlycreative@gmail.com                   NaN   
2733               mlonghelt@gmail.com                   NaN   

                     cityName         category  \
0  

## Meaning of the columns
https://developers.facebook.com/docs/instagram-api/reference/ig-user/
This URL has field defintion.
* biography - Profile bio text.
* followers_count - Total number of Instagram users following the user.
* following_count - Total number of Instagram users the user follows.
* media_count - Total number of IG Media published on the user.


In [6]:
set(df_women_event_mined.username)

0        hannahmariethemuse
1            refinedlash.co
2         studio_irene_rose
3              alena_smily9
4        sincerelyluna.xoxo
                ...        
11276                lianev
11277                   kuz
11278           tanamongeau
11279             lilyachty
11280             jenselter
Name: username, Length: 11281, dtype: object

In [12]:
df_women_event

username  media_count                      full_name  \
0             ariannyceleste         6039       Arianny Celeste UFC Mama   
1                 sarahstage         1501                    Sarah Stage   
2     beauty_nurse_elizabeth         1060           Elizabeth Sommers RN   
3                  carlyrbel          493                      CARLY BEL   
4              houdaschimmel          507                 Houda Schimmel   
...                      ...          ...                            ...   
1467            houghtonmade            1  Memorable and Detailed Events   
1468               violacomm            3           Viola Communications   
1469  brainstorm.media.group            0         Brainstorm Media Group   
1470     littlestudio_beauty           33           Little Studio_Beauty   
1471           adtop1percent            0   Ann Dee | LA ENTREPRENEUR LA   

      following_count  follower_count first_name  \
0                 997         3288148    Celeste   
1                1472         2086691      Sarah   
2                 383         1504557  Elizabeth   
3                 191         1035410      Carly   
4                3041          693036      Houda   
...               ...             ...        ...   
1467              166              21     Lauren   
1468                3              20      Viola   
1469               32              20  Katherine   
1470              146              12   Laetitia   
1471               42               7        Ann   

                          external_url  engagement_percent  \
0     http://www.ariannyexclusive.com/                0.06   
1         https://linktr.ee/Sarahstage                0.01   
2     http://www.elizabethsommers.com/                0.23   
3                 http://carlybel.com/                0.14   
4                                  NaN                 NaN   
...                                ...                 ...   
1467                               NaN               10.95   
1468   http://violacommunications.com/                6.83   
1469  http://brainstormmediagroup.com/                0.00   
1470       http://www.littlestudio.us/                4.03   
1471                               NaN                0.00   

                                   email contact_phone_number  \
0             arianny@ariannyceleste.com                  NaN   
1               booksarahstage@gmail.com                  NaN   
2                    sommersrn@gmail.com                  NaN   
3                carlybelsnaps@gmail.com                  NaN   
4               infos@shimmypetstore.com                  NaN   
...                                  ...                  ...   
1467             laurenjmadrid@gmail.com         +13234008726   
1468        info@violacommunications.com                  NaN   
1469  katherine@brainstormmediagroup.com                  NaN   
1470            laetitia@littlestudio.fr                  NaN   
1471             adtop1percent@gmail.com                  NaN   

                     cityName       category  \
0     Los Angeles, California  Public figure   
1                         NaN            NaN   
2     Los Angeles, California  Health/beauty   
3                         NaN   Entrepreneur   
4     Los Angeles, California   Entrepreneur   
...                       ...            ...   
1467  Los Angeles, California            NaN   
1468           Tampa, Florida            NaN   
1469  Los Angeles, California            NaN   
1470                      NaN            NaN   
1471                      NaN   Entrepreneur   

                                              biography address_street  \
0     Raidens Mom 👶🏽🐶🐶🧿 Founder @girlfriendbox @them...            NaN   
1     Boy Mom 👶🏻👦🏻 Wife 💃🏻 & Online Fitness Coach 💪🏼...            NaN   
2     Aesthetic RN-BSN •Injection Artist 🌹 •Natural ...            NaN   
3     One of a kind ✨ • @carlybelx • shop @carlyclub...            NaN   
4        

In [9]:
len(set(df_women_event_mined.username).intersection(set(df_women_event.username)))

3

In [62]:
df_buy_biz

username Openness Conscientiousness Extroversion  \
0               goodeatzco      Low              High         High   
1                lexxalynn      NaN               NaN          NaN   
2               a1republic     High              High         High   
3           keychainsocial     High              High         High   
4        thegiftshoppe.nft    Vague               Low          Low   
..                     ...      ...               ...          ...   
138           irose.social     High              High         High   
139               nattcity     High               Low         High   
140             jondennill     High              High         High   
141            igmodels_co      Low              High        Vague   
142  skystrategicmarketing     High              High         High   

    Agreeableness Neuroticism      Type  media_count                full_name  \
0           Vague       Vague      Blog         1693       #GoodEatzCo | Andy   
1             NaN         NaN       NaN          241          Alexa Ditchburn   
2            High         Low  Business           30            A1 REPUBLIC ™   
3            High         Low  Business          214    Keychain Social, Inc.   
4           Vague       Vague  Business            9          the gift shoppe   
..            ...         ...       ...          ...                      ...   
138           Low         Low  Business          338                    IROSE   
139           Low         Low  Personal         1129                 nathalie   
140          High         Low  Personal          124         Jonathan Dennill   
141         Vague       Vague      Blog          458    IG Models Worldwide ®   
142          High         Low  Business          427  Sky Strategic Marketing   

     following_count  ...  engagement_percent  \
0                145  ...                0.14   
1                999  ...                0.00   
2                613  ...                0.53   
3                  0  ...                 NaN   
4                  5  ...                0.57   
..               ...  ...                 ...   
138              422  ...                0.30   
139              686  ...                0.06   
140              733  ...                0.49   
141              158  ...                0.01   
142             2133  ...                0.18   

                               email contact_phone_number  \
0               goodeatzco@gmail.com                  NaN   
1          alexadesigns312@gmail.com                  NaN   
2               a1republic@yahoo.com         1.949934e+10   
3            info@keychainsocial.com                  NaN   
4        hello@fromthegiftshoppe.com                  NaN   
..                               ...                  ...   
138           sharon@irosesocial.com                  NaN   
139  nathalia@nfluenceconsulting.com                  NaN   
140            jon.dennill@gmail.com                  NaN   
141            igmodels.co@gmail.com                  NaN   
142  tanya@skystrategicmarketing.com         1.813337e+10   

                      cityName             category  \
0      Los Angeles, California              Blogger   
1                          NaN                  NaN   
2    Newport Beach, California  Social Media Agency   
3      Los Angeles, California                  NaN   
4                          NaN                  NaN   
..                         ...                  ...   
138    Los Angeles, California      Product/service   
139                        NaN                Actor   
140                        NaN                  NaN   
141    Los Angeles, California  Social Media Agency   
142             Tampa, Florida                  NaN   

                                             biography  \
0    🍴Professional Food Consumer🤘🏻😛 Tiktok 🎥: @good...   
1    Web & Graphic Designer👩🏽‍💻 🧶Etsy: Knit Wits by...   
2    The #1 organic IG growth service for influ

In [9]:
## try out scikit learn's  select k best class

import pandas as pd
import numpy as np
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn import preprocessing 

def calculate_feature_importance(data, trait):
    columns = ['Type','media_count','following_count', 'follower_count', 'engagement_percent', 'AvgLikes', 'AvgComments', trait]
    X = data[columns]  #independent columns
    X = X.dropna()
    label_encoder = preprocessing.LabelEncoder()
    X[trait] = label_encoder.fit_transform(X[trait])
    X['Type'] = label_encoder.fit_transform(X['Type'])

    y = X[[trait]]
    X.drop(columns=trait, inplace=True)


    #apply SelectKBest class to extract top 10 best features
    bestfeatures = SelectKBest(score_func=chi2, k='all')
    fit = bestfeatures.fit(X,y)
    dfscores = pd.DataFrame(fit.scores_)
    dfcolumns = pd.DataFrame(X.columns)
    #concat two dataframes for better visualization 
    featureScores = pd.concat([dfcolumns,dfscores],axis=1)
    featureScores.columns = ['Specs','Score']  #naming the dataframe columns
    print(featureScores.nlargest(10,'Score'))  #print 10 best features

In [10]:
import pandas as pd
feature_columns = ['username', 'Type','media_count','following_count', 'follower_count', 'engagement_percent', 'AvgLikes', 'AvgComments']
all_five_traits = ['Openness', 'Conscientiousness','Extroversion', 'Agreeableness', 'Neuroticism']

all_df = pd.DataFrame()

for df in [df_buy_insta_accounts, df_buy_biz, df_buy_insta_accounts_add_on]:
    df_focus = df[df['Openness'].notna()]
    df_focus = df_focus[feature_columns+all_five_traits]
    all_df = pd.concat([all_df, df_focus])


In [12]:

for trait in all_five_traits:
    print("=========")
    print(trait)
    calculate_feature_importance(all_df, trait)


Openness
                Specs          Score
3      follower_count  407487.936396
2     following_count    7829.749073
1         media_count    3974.455974
5            AvgLikes    1379.433089
6         AvgComments     281.594431
0                Type       6.496976
4  engagement_percent       5.344649
Conscientiousness
                Specs         Score
3      follower_count  86472.967375
5            AvgLikes  43160.963575
2     following_count  14138.324108
1         media_count   4148.089570
6         AvgComments    242.680015
4  engagement_percent     85.870360
0                Type      2.375934
Extroversion
                Specs          Score
3      follower_count  108297.689778
1         media_count   25191.254344
5            AvgLikes   14515.187110
2     following_count    3418.331942
6         AvgComments     142.745645
4  engagement_percent      25.355317
0                Type      12.604089
Agreeableness
                Specs          Score
3      follower_count  118696

## Call Instagram Business Discovery API

`df_buy_insta_accounts` has 94 rows with big 5 score, total 5535 rows

`df_buy_biz` total 143 rows, 129 rows have big 5 scores

`df_buy_insta_accounts_add_on` 2734 rows, 95 have big 5 scores

=======

`df_women_event_mined` Big 5 column is empty, total 11281 rows


`df_women_event` 1472 rows, no big five scores

In [6]:
import requests
import json
import pandas as pd
def get_media_metadata(ig_user_name, token, pages=1):
    params = {
        "ig_username": ig_user_name,
        "access_token": token,
        "endpoint_base":"https://graph.facebook.com/v17.0/",
        "instagram_account_id":"17841400224190654"
    }
    biz_discovery_first_page = '){username,ig_id,id,profile_picture_url,biography,follows_count,followers_count,media_count, media{comments_count, like_count, caption, timestamp, media_type}}' 

    endpointParams = dict() # parameter to send to the endpoint
    
    endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_first_page
    
    # string of fields to get back with the request for the account
    endpointParams['access_token'] = params['access_token'] # access token

    url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
    
    data = requests.get(url, endpointParams)
    if data.ok:
        dict_ = json.loads(data.content)
    else:
        return data.reason
    try:
        df = pd.DataFrame(dict_['business_discovery']['media']['data'])
    except:
        print(f"{ig_user_name} has bad data")
        return dict_
    
    pages -=1
    while pages>0:
        #d1 = json.loads(data.content)
        try:
            cursor = dict_['business_discovery']['media']['paging']['cursors']['after']
        except:
            return "no cursor"
        biz_discovery_following_page = f"){{username,website,name,ig_id,id,profile_picture_url,biography,follows_count,followers_count,media_count, media.after({cursor}){{comments_count, like_count, caption, timestamp, media_type}}}}"
        endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_following_page
        url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
        data = requests.get(url, endpointParams)
        dict_ = json.loads(data.content)
        df = pd.concat([df, pd.DataFrame(dict_['business_discovery']['media']['data'])])
                       
        pages -=1
        
    
    return df


In [30]:
from tqdm import tqdm
import os
from pathlib import Path

# go to: https://developers.facebook.com/tools/explorer/
# to get token
# this token is good till August 17th
token ="EAACccrMmWa0BAGziHtFogA885lMeFU3LHBHXDqZBtAMiTdmCUw7WLprdQUMMWKFZCjqhnlcDhZC2nIYFPzohpRsLBaaTopC4n2FAmqwbrTbJZBTixRBEhBlrrhl80tF3apZC2wEEWaj6bzVKKMepM4990lzugk5kWZB8egWVc8ZAgZDZD"

def get_biz_discovery_e2e(usernames, folder):
    
    try:
        os.mkdir(f"feature/{folder}")
    except:
        pass

    scraped_users = []
    for entry in Path(f"feature/{folder}").iterdir():
        # check if it a file
        if entry.is_file():
            scraped_users.append(entry.stem)

    bad_users = []
    for ig_username in tqdm(usernames.tolist()):
        if ig_username in scraped_users:
            continue

        res = get_media_metadata(ig_username, token, 1)
        if type(res) is str:
            if res == "Forbidden":
                print("possibly reach rate limit")
                break
            bad_users.append({ig_username: res})
        else:
            res.to_parquet(f"feature/{folder}/{ig_username}.parquet")
    return bad_users, scraped_users

In [31]:
usernames = df_buy_insta_accounts[df_buy_insta_accounts['Openness'].isna()]['username'][300:500]
folder = "buy_insta_accounts"
bad_users, scraped_users = get_biz_discovery_e2e(usernames, folder)

 98%|█████████▊| 195/200 [00:04<00:00, 45.75it/s] 

possibly reach rate limit


### get bio only

In [36]:
import requests
import json
import pandas as pd
from tqdm import tqdm
import os
from pathlib import Path
import pickle

def get_ig_user_bio(ig_user_name, token):
    params = {
        "ig_username": ig_user_name,
        "access_token": token,
        "endpoint_base":"https://graph.facebook.com/v17.0/",
        "instagram_account_id":"17841400224190654"
    }
    biz_discovery_first_page = '){username,ig_id,id,biography}' 

    endpointParams = dict() # parameter to send to the endpoint
    
    endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_first_page
    
    # string of fields to get back with the request for the account
    endpointParams['access_token'] = params['access_token'] # access token

    url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
    
    data = requests.get(url, endpointParams)
    if data.ok:
        dict_ = json.loads(data.content)
    else:
        return data
    return dict_

def get_ig_user_bio_e2e(usernames, folder):
    try:
        os.mkdir(f"feature/{folder}")
    except:
        pass

    scraped_users = []
    for entry in Path(f"feature/{folder}").iterdir():
        # check if it a file
        if entry.is_file():
            scraped_users.append(entry.stem)

    bad_users = []
    for ig_username in tqdm(usernames.tolist()):
        if ig_username in scraped_users:
            continue

        res = get_ig_user_bio(ig_username, token)
        if type(res) is str:
            if res == "Forbidden":
                print("possibly reach rate limit")
                break
            bad_users.append({ig_username: res})
        else:
            with open(f"feature/{folder}/{ig_username}.pkl",'wb') as f_out:
                pickle.dump(res, f_out)
    return bad_users, scraped_users

In [37]:
token ="EAACccrMmWa0BO9iX5gPUQW97WI9CMkzHtq6VyYlqcVKzZCFkJj2N36QYQIQWsYRoLo2spZBBXpPyUuInZAD29j8ZCLj2yMXva3PftTzCa90llgJK9rxPTJ38xhHU181H7G65x0eKicwibPovw377chZBybpQrn5Bp84vjIKQR0fspjCBYx7yZB8Oj4SWHeKTsPcAcSj3GqFw1E8JCIoq8mZBi5D"

res = get_ig_user_bio("wendyalane1", token)

In [20]:
res

{'business_discovery': {'username': 'wendyalane1',
  'ig_id': 189742646,
  'id': '17841401399382964',
  'biography': '⭕️ Founder of @hollywoodwinnerscircle & \n🎭Talent Managers For Actors on FB\n🎬 #1 Acting Business Coach\n🖥 YouTube (6M+ Views)\nFREE Masterclass⬇️'},
 'id': '17841400224190654'}

In [38]:
usernames = df_buy_insta_accounts[df_buy_insta_accounts['Openness'].isna()]['username'][:10]
folder = "buy_insta_accounts"
bad_users, scraped_users = get_ig_user_bio_e2e(usernames, folder)

100%|██████████| 10/10 [00:02<00:00,  3.53it/s]


In [11]:
#combine 
import pandas as pd
from pathlib import Path

p = Path('feature')
folders = [x for x in p.iterdir() if x.is_dir() and x.stem.startswith("buy")]

full_df = pd.DataFrame()
for data_dir in folders:
    for parquet_file in data_dir.glob('*.parquet'):
        ig_username = parquet_file.stem
        df = pd.read_parquet(parquet_file)
        df['ig_username'] = ig_username
        full_df = pd.concat([full_df, df])

In [12]:
import pandas as pd
full_df = pd.read_parquet("concat_feature.parquet")

In [23]:
full_df.ig_username.sample()

3    biancabosch
Name: ig_username, dtype: object

In [27]:
full_df[full_df.ig_username=='wendyalane1']

comments_count  like_count  \
0                5        88.0   
1               20       400.0   
2               14       243.0   
3                9       341.0   
4                6       157.0   
5               21       483.0   
6                0        82.0   
7                0        32.0   
8               25       417.0   
9                1        49.0   
10              12       113.0   
11               5        87.0   
12               9       218.0   
13               2        83.0   
14               6       182.0   
15               5       124.0   
16              50       613.0   
17               9       126.0   
18               4        54.0   
19              45       613.0   
20               0        65.0   
21              80      1288.0   
22               6       135.0   
23              21       209.0   
24               8       105.0   

                                              caption  \
0   Several years ago Urusha Pandey watched my vid...   
1   Actors do you want to be successful in this bu...   
2   The time you spend doubting yourself and wonde...   
3   Actors One of the biggest mistakes actors make...   
4           You are here to give and serve the world.   
5   This book is amazing! So moving, inspiring and...   
6                                                None   
7   What I teach you in Hollywood Winners Circle w...   
8   You are working hard. You’re focused on your g...   
9   Actors, I am a Talent Manager. I know the insi...   
10  One of the best weeks of my life with my dear ...   
11                               Photo by Amber Blake   
12  Understand spending the day with one of my fav...   
13                                               None   
14  Every single day, we get brand new actors on t...   
15  Ooooooh!!!! Previews…. And THEN!!! 🧜🏾‍♀️ 🧜🏾‍♀️...   
16  It takes a lot of courage to put yourself out ...   
17  Kim, we are so darn proud of you. You’re talen...   
18  ANOTHER booking from ERICK MCALLISTER. HWC wor...   
19                   I pour into people I believe in.   
20  Highly Recommend!\nFILM MAMKING CLASS\n\nJim D...   
21  Skin is just a coating. If you can’t understan...   
22  ACTORS Are you ready to join the HWC family? \...   
23  My man….\nOkay, \nHe must be KIND (to me and e...   
24  💥 Ready to TRANSFORM your acting career? 🌟 The...   

                   timestamp      media_type                 id  ig_username  \
0   2023-06-20T07:03:33+0000  CAROUSEL_ALBUM  18029365417481034  wendyalane1   
1   2023-06-19T08:26:02+0000           VIDEO  17911618985691215  wendyalane1   
2   2023-06-18T20:52:08+0000           VIDEO  17889071324792756  wendyalane1   
3   2023-06-17T07:00:35+0000           VIDEO  17909572247696743  wendyalane1   
4   2023-06-16T08:29:03+0000           IMAGE  17990166371082372  wendyalane1   
5   2023-06-16T02:29:48+0000           IMAGE  17847809537995993  wendyalane1   
6   2023-06-16T02:16:22+0000           VIDEO  17958046562558841  wendyalane1   
7   2023-06-14T20:27:28+0000           VIDEO  17968187252291358  wendyalane1   
8   2023-06-14T20:15:15+0000           VIDEO  18213336712218167  wendyalane1   
9   2023-06-14T01:54:01+0000  CAROUSEL_ALBUM  18027723079483127  wendyalane1   
10  2023-06-12T08:45:03+0000           VIDEO  18251108740088052  wendyalane1   
11  2023-06-10T04:58:18+0000           IMAGE  17984221994177734  wendyalane1   
12  2023-06-10T04:43:10+0000  CAROUSEL_ALBUM  18192270925267831  wendyalane1   
13  2023-06-10T00:07:46+0000  CAROUSEL_ALBUM  17982986858281740  wendyalane1   
14  2023-06-09T10:45:43+0000  CAROUSEL_ALBUM  17992530364848759  wendyalane1   
15  2023-06-09T07:35:19+0000  CAROUSEL_ALBUM  18026420668492928  wendyalane1   
16  2023-06-08T22:42:29+0000           IMAGE  17975227967339915  wendyalane1   
17  2023-06-08T19:58:27+0000  CAROUSEL_ALBUM  18021090622520917  wendyalane1   
18  2023-06-04T02:57:33+0000           IMAGE  17973244862464698  wendyalane1   
19  2023-06-03T00:51:25+0000        

In [2]:
# get bad_user json stats
import json
from pathlib import Path
user_reason = []
p = Path('jobs/feature')
folders = [x for x in p.iterdir() if x.is_dir()]

for data_dir in folders:
    for bad_data_json in data_dir.glob("*.json"):
        print(bad_data_json)
        with open(bad_data_json) as f_in:
            data = f_in.read()
            data=data.replace('][',',')
            j = json.loads(data)
            user_reason.extend(j)

jobs/feature/df_buy_insta_accounts_add_on/bad_users_idx.json
jobs/feature/df_buy_insta_accounts_add_on/bad_users_idx_0.json
jobs/feature/women_event_mined/bad_users_idx.json
jobs/feature/women_event/bad_users_idx.json
jobs/feature/buy_insta_accounts/bad_users_idx.json
jobs/feature/buy_biz/bad_users_idx.json


In [31]:
# check unique bad users:
users_s = set()
for user in user_reason:
    users_s.add([*user.keys()][0])

In [3]:
user_reason[100]

{'sheba.bonita': 'Bad Request'}

In [39]:
bad_users = users_s.difference(full_df.index)

In [54]:
# what are the reasons for these base users?
from collections import defaultdict
user_reason_real = []
str_reason_user = defaultdict(set)
non_str_reason_user = {}
for user in user_reason:
    ig_username = [*user.keys()][0]
    if ig_username not in bad_users:
        continue
    if type(user[ig_username]) is str:
        str_reason_user[user[ig_username]].add(ig_username)
    else:
        non_str_reason_user[ig_username]=user

In [60]:
len(str_reason_user['Bad Request'])

6442

## Preprocess Features

In [3]:
from emot.emo_unicode import UNICODE_EMO, EMOTICONS
import re

def lower_text(text):
    try:
        text = text.lower()
    except:
        return ""
    return text
def remove_newline(string):
    string = string.replace("\n", "")
    return string
def remove_punctuation(string): 
    punctuations = '''()-[]{};:'"\<>/@#$%^&*~'''
    for x in string.lower(): 
        if x in punctuations: 
            string = string.replace(x, "")
        elif x == ",._":
            string = string.replace(x, " ")
        elif x == "?":
            string = string.replace(x, "[question mark]")
    return string

def convert_emojis(text):
    for emot in UNICODE_EMO:
        text = text.replace(emot, "_".join(UNICODE_EMO[emot].replace(",","").replace(":","").split()))
    return text

def convert_emoticons(text):
    for emot in EMOTICONS:
        text = re.sub(u'('+emot+')', "_".join(EMOTICONS[emot].replace(",","").split()), text)
    return text

def combine_metadata(row):
    return f"caption {row['caption']}, is of type {row['media_type'].lower()}, has {row['comments_count']} comments and {row['like_count']} likes"


In [4]:
full_df['full_caption'] = full_df.apply(combine_metadata, axis=1)
print("lower text")
full_df['full_caption'] = full_df['full_caption'].apply(lower_text)
print("remove newline")
full_df['full_caption'] = full_df['full_caption'].apply(remove_newline)
print("convert emojos")
full_df['full_caption'] = full_df['full_caption'].apply(convert_emojis)
print("convert emoticons")
full_df['full_caption'] = full_df['full_caption'].apply(convert_emoticons)
full_df_concat = full_df[['ig_username','full_caption']].groupby("ig_username").sum()

lower text
remove newline
convert emojos
convert emoticons


In [1]:
import pandas as pd
full_df = pd.read_parquet("prediction_feature.parquet")

In [11]:
full_df.full_caption.iloc[0]

'caption out now on all platformsglobe_showing_Asia-Australia, is of type carousel_album, has 11 comments and 824.0 likescaption black truck szn.black_large_square️, is of type carousel_album, has 12 comments and 2593.0 likescaption link in my bio\U0001faf5medium-light_skin_tonegon tell a friendperson_golfingmedium_skin_tone\u200dmale_sign️, is of type video, has 26 comments and 2241.0 likescaption in another n!gga city🥸dos x tmbpgstrapp, is of type carousel_album, has 26 comments and 3934.0 likescaption double’o 4 da people., is of type image, has 11 comments and 1615.0 likescaption my dawg pgstrapp bringing me out to his city i need everyone yo show tf out ofng\U0001faf5medium-light_skin_tone🧨any ticket inquiries dm my boy pgstrapp lets do it do itperson_golfingmedium_skin_tone\u200dmale_sign️, is of type video, has 21 comments and 1298.0 likescaption you know i gotta show love to my ricansface_with_tears_of_joy\U0001faf6medium-light_skin_tone 🇵🇷, is of type carousel_album, has 75 co

In [5]:
full_df_concat.to_parquet("prediction_feature.parquet")

In [4]:
full_df['full_caption_count']=full_df['full_caption'].apply(lambda s: len([*s.split(" ")]))

## These are used particularly for training data prep

In [16]:
columns = ['username', 'Openness','Conscientiousness','Extroversion','Agreeableness','Neuroticism']
feature_df =full_df_concat.merge(gc_df, left_index=True, right_on="username", how="left")
gc_df = pd.concat([
    df_buy_insta_accounts[columns],
    df_buy_biz[columns],
    df_buy_insta_accounts_add_on[columns]
])

NameError: name 'gc_df' is not defined

In [51]:
feature_df

full_caption          username  \
59    caption did you know henry ford made the 8 hou...            36mktg   
69    caption so clean smiling_face_with_heart-eyesr...    6staroileaters   
26    caption through the night crescent_moon • maet...        740project   
57    caption take your social media accounts out of...  _boostfruit.com_   
10    caption appreciate you excellencealways folded...      _cloutfinder   
...                                                 ...               ...   
71    caption none, is of type image, has 1 comments...      wipmarketing   
14    caption enjoy this video while i disappear for...         xcarmeeen   
5461  caption power back.#fanggang | xfl, is of type...         xflvipers   
67    caption summeeeeer! all the hard work of clean...       yogabody_uk   
84    caption now booking grammy award winning artis...          zgroupla   

      Openness  Conscientiousness  Extroversion  Agreeableness  Neuroticism  
59         1.0                1.0           1.0            1.0          0.0  
69         0.0                1.0           0.5            0.5          0.5  
26         1.0                0.0           0.0            0.0          0.5  
57         1.0                0.0           1.0            0.5          0.5  
10         1.0                0.0           1.0            0.5          0.5  
...        ...                ...           ...            ...          ...  
71         1.0                1.0           1.0            1.0          0.0  
14         0.0                0.0           1.0            0.0          1.0  
5461       0.0                1.0           1.0            1.0          0.0  
67         1.0                1.0           1.0            1.0          0.0  
84         1.0                0.0           1.0            0.5          0.5  

[269 rows x 7 columns]

In [50]:
feature_df['Openness'] = feature_df['Openness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Conscientiousness'] = feature_df['Conscientiousness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Extroversion'] = feature_df['Extroversion'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Agreeableness'] = feature_df['Agreeableness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Neuroticism'] = feature_df['Neuroticism'].map({"High":1.0,"Low":0,"Vague":0.5})

In [20]:
feature_so_far = feature_so_far.dropna()

NameError: name 'feature_so_far' is not defined

In [52]:
feature_df.to_parquet("feature.parquet")

In [37]:
from google.cloud import storage

client = storage.Client()
bucket = client.get_bucket('feature')
    

Forbidden: 403 GET https://storage.googleapis.com/storage/v1/b/feature?projection=noAcl&prettyPrint=false: 100822155188-compute@developer.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket. Permission 'storage.buckets.get' denied on resource (or it may not exist).